mllib/src/main/scala/org/apache/spark/mllib/clustering/KMeans.scala

```diff
@@ -290,29 +311,30 @@ class KMeans private (
       val bcActiveCenters = sc.broadcast(activeCenters)
 
       // Find the sum and count of points mapping to each center
-      val totalContribs = data.mapPartitions { points =>
-        val thisActiveCenters = bcActiveCenters.value
-        val runs = thisActiveCenters.length
-        val k = thisActiveCenters(0).length
-        val dims = thisActiveCenters(0)(0).vector.size
-
-        val sums = Array.fill(runs, k)(Vectors.zeros(dims))
-        val counts = Array.fill(runs, k)(0L)
-
-        points.foreach { point =>
-          (0 until runs).foreach { i =>
-            val (bestCenter, cost) = KMeans.findClosest(thisActiveCenters(i), point)
-            costAccums(i) += cost
-            val sum = sums(i)(bestCenter)
-            axpy(1.0, point.vector, sum)
-            counts(i)(bestCenter) += 1
+      val totalContribs = data.sample(false, miniBatchFraction, 42 + iteration)
```

Comment:
This change is pretty reasonable. I think Derrick Burns had suggested something like this a long time ago, too. It needs to be brought up to date after the 'runs' changes. We also need to make sure this step here doesn't add any overhead when the fraction is 1.